### ISSUE TITLE
Selecting text will copy it into the pasteboard

### ISSUE THREAD

**Issue by kkk669** [tbodt/ish#174]
In the latest TestFlight build 35, when I just long press text to select it on iSH, it will be copied into the pasteboard even if I don’t tap the “Copy” item.

How to reproduce it: https://youtu.be/OoDOArAkhSg

1. Select text
2. Tap “Paste” item

I reproduced the issue on

- 3rd generation 12.9” iPad Pro with iOS 12.1.1
- iPad Air 2 with iOS 12.1
- iPhone XS with iOS 12.1.1

Related to #25

**Comment by ELChris414** [comment]
I can consistently not reproduce this with video proof 😅

**Comment by tbodt** [comment]
I can’t reproduce either, on 11.4. This is definitely not supposed to happen though.

**Comment by kkk669** [comment]
@ELChris414 @tbodt That’s weird. 🤔

The issue doesn’t happen on other apps such as Notes, Safari, and Blink Shell.

**Comment by francoisvignon** [comment]
I confirm the issue (ios 12.1.1)
and so, I can't past text from other sources.

**Comment by kkk669** [comment]
@francoisvignon If you use an external keyboard, you can paste by pressing CMD+V. I don’t know any other way.

**Comment by francoisvignon** [comment]
many thanks for your answer …. no external keyboard available :-/

why not using Ctrl+V as Ctrl is already present ?

____________________

Francois

**Comment by itskenny0** [comment]
You could try using SwiftKey for now. It has a clipboard manager and can paste without app support.

https://m.imgur.com/gallery/keK5ysH

**Comment by tbodt** [comment]
@francoisvignon ctrl-v enters visual block mode in vim.

**Comment by francoisvignon** [comment]
ok :-/ ... will wait for long past not longer direct copy to clipboard ;-)

send from my mobile.
____________________

François

**Comment by tbodt** [comment]
I've updated to 12.1.1 and still can't reproduce on iPod Touch.

**Comment by Patronics** [comment]
I'm having this issue too on iPhone X with iOS 12.1, perhaps as a workaround (especially since you seem unable to reproduce it @tbodt, and is presumably therefore pretty hard to debug), you could add a dedicated paste key in the row of modifier keys? For me that would also be more intuitive to use, and presumably avoid this bug (based on those saying that external keyboards with Cmd + v work fine)

**Comment by francoisvignon** [comment]
https://share.icloud.com/photos/02EaRk5ZN7qKJsQXqfifBCdBQ

a small how to reproduce
long click on ´tutu' :
the text is select and copy/paster is show, then click on copy

long click at cursor level:
a part of the line is selected and  cooy/paste is show, then click on paste

the pasted text is the last selected, not the last copied.

I hope it's helpfull


send from my mobile.
____________________

François

**Comment by francoisvignon** [comment]
maybe to help: once text is selected and without pressing copy, the selected text is in the clipboard: selecting text in iSh (with button copy|paste still displayed) and pasting on the Mac past the selected text.

shortly, it’s seem like  (void)copy:(id)sender was called when dragging the selection …. and after some try, the 'copy'  was called when stoping selection, when copy|paste button appear.

I don’t know enough Xcode and iOs development to be more helpful at this time :-/ 

PS: I know copy/paste is not an important feature, but it’s helpful to copy/past shell command and other source as long Files App cannot write iSh files ;-)

_____________
Francois

**Comment by francoisvignon** [comment]
I don't know .... if it was under windows (arggghh sorry), it can be mistake between mouse up and copy to clipboard (maybe not the most accurate sample, but that's all I have at this time).

maybe a wrong link (or other) in the storyboard invoking copy when release the hang ... no more idea.

but the computer is the truth: if something is done, it's because somebody do it.

So, as other app don't have the same behavior, I think we can exclude iOs and the hardware.

the source (!) of the problem is, I think (sorry for that, but I know), between the chair and the keyboard.

Sometime, solving small bug need a lot of effort :-(

send from my mobile.
____________________

François

**Comment by Patronics** [comment]
Yeah, I've had the same observation, selecting text copies it, regardless of using or not the copy/paste dialog

**Comment by francoisvignon** [comment]
maybe the problem is in the js related to term ?

send from my mobile.
____________________

François

**Comment by francoisvignon** [comment]
I have traced (void)copy:(id)sender and this one, when long press and selecting is called a lot (6 to 8 time) without tapping copy, and is called one more time when tapping copy.

the first lot of call was performed with a null sender (don't ask me why, I'm not an iOs UI guru) and the one when tapping copy is called with a UIMenuController sender.

so, I have simply modify (void)copy:(id)sender with the following:
```
(void)copy:(id)sender {
    if(sender != NULL)
    {
        // not longer necessary NSLog(@"copy called %@", sender);

     [self.terminal.webView evaluateJavaScript:@"window.getSelection().toString()" completionHandler:^(NSString *selection, NSError *err) {
        if (err != nil) {
            NSLog(@"error copying: %@", err);
            return;
        }
        UIPasteboard.generalPasteboard.string = selection;
    }];
    }
}

```
and, tada :-), long press and selecting no longer copy, but now you should press copy to copy to the clipboard ;-)

a side effect is a long tap select and show the paste button, this one paste the previous copied text (like the paste button above the keyboard .... )

**Comment by tbodt** [comment]
@

**Comment by tbodt** [comment]
@francoisvignon thanks for debugging! I’m curious what the stack trace is when copy: is called with a nil sender. You can find it in the debug panel on the left sidebar, or by typing `backtrace` into LLDB. If you use the debug panel, you’ll want to click the tiny button at the bottom that enables showing stack frames that are part of system libraries, otherwise you probably won’t see anything useful.

**Comment by francoisvignon** [comment]
the thread stack I seen is not helpfull: I seen copy is called this way
	UIApplicationMain
		NSObject_QSSupport _accessibilityQuickSpeakContent
			UIDocumentSharingController
				break point in copy

I will try according your indication tomorrow ...

**Comment by tbodt** [comment]
> NSObject_QSSupport _accessibilityQuickSpeakContent

wat

that could be very helpful actually

**Comment by francoisvignon** [comment]
"wat" ?

**Comment by tbodt** [comment]
I can reproduce this if Settings > Accessibility > Speech > Speak Text is on.

**Comment by francoisvignon** [comment]
yesss !!!! so, now, as you can reproduce it, you can solve it.

send from my mobile.
____________________

François

**Comment by francoisvignon** [comment]
now we can use copy/past, we have another issue:
I have typed uname -a, ok
after I have selected "uname -a" and pasted, and then an error: uname -a not found.
so I have dump what I have copied:
echo " (then past) " .... and you can see the result bellow.
I think it's a problem in text encoding.


iFv:~# uname -a                                  
Linux iFv 3.2.0-ish SUPER AWESOME compiled on Dec
 29 2018 i686 Linux                              
iFv:~# uname -a                                  
-ash: uname -a: not found                        
iFv:~# echo "uname -a" | od -xc                  
0000000   u   n   a   m   e 302 240   -   a  \n  
        6e75 6d61 c265 2da0 0a61                 
0000012                                          
iFv:~#

**Comment by francoisvignon** [comment]
sorry for the bad presentation :-/ markdown

**Comment by francoisvignon** [comment]
the problem is the space is pasted as an unicode non breakable space (\302 \240) and not as a space (32)

**Comment by francoisvignon** [comment]
thx


_____________
Francois

**Comment by francoisvignon** [comment]
and you are one step forward on the path to be an UI Kit Guru ;-)

send from my mobile.
____________________

François
